thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/onboard/pinot/metrics/OnboardPinotMetricsService.java

```diff
@@ -0,0 +1,323 @@
+import com.linkedin.thirdeye.datalayer.bao.DatasetConfigManager;
+import com.linkedin.thirdeye.datalayer.bao.IngraphMetricConfigManager;
+import com.linkedin.thirdeye.datalayer.bao.MetricConfigManager;
+import com.linkedin.thirdeye.datalayer.dto.DashboardConfigDTO;
+import com.linkedin.thirdeye.datalayer.dto.DatasetConfigDTO;
+import com.linkedin.thirdeye.datalayer.dto.IngraphMetricConfigDTO;
+import com.linkedin.thirdeye.datalayer.dto.MetricConfigDTO;
+import com.linkedin.thirdeye.datalayer.pojo.DashboardConfigBean;
+
+/**
+ * This is a service to onboard datasets automatically to thirdeye from pinot
+ * This service runs periodically and checks for new tables in pinot, to add to thirdeye
+ * If the table is an ingraph table, it loads metrics from the ingraph table
+ * It also looks for any changes in dimensions or metrics to the existing tables
+ */
+public class OnboardPinotMetricsService implements Runnable {
+  private static final Logger LOG = LoggerFactory.getLogger(OnboardPinotMetricsService.class);
+
+
+  private static final DAORegistry DAO_REGISTRY = DAORegistry.getInstance();
+
+  private DatasetConfigManager datasetConfigDAO = DAO_REGISTRY.getDatasetConfigDAO();
+  private MetricConfigManager metricConfigDAO = DAO_REGISTRY.getMetricConfigDAO();
+  private DashboardConfigManager dashboardConfigDAO = DAO_REGISTRY.getDashboardConfigDAO();
+  private IngraphMetricConfigManager ingraphMetricConfigDAO = DAO_REGISTRY.getIngraphMetricConfigDAO();
+
+
+  private CacheResource cacheResource;
+
+  private ScheduledExecutorService scheduledExecutorService;
+  private OnboardPinotMetricsUtils onboardPinotMetricsUtils;
+
+  private List<String> allDatasets = new ArrayList<>();
+  private Map<String, Schema> allSchemas = new HashMap<>();
+
+  public OnboardPinotMetricsService() {
+  }
+
+  public OnboardPinotMetricsService(ThirdEyeConfiguration config) {
+
+    onboardPinotMetricsUtils = new OnboardPinotMetricsUtils(config);
+    scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+    cacheResource = new CacheResource();
+  }
+
+  public void start() {
+    scheduledExecutorService.scheduleAtFixedRate(this, 0, 4, TimeUnit.HOURS);
+  }
+
+  public void shutdown() {
+    scheduledExecutorService.shutdown();
+  }
+
+  public void run() {
+    try {
+      loadDatasets();
+
+      for (String dataset : allDatasets) {
+        LOG.info("Checking dataset {}", dataset);
+        DatasetConfigDTO datasetConfig = datasetConfigDAO.findByDataset(dataset);
+        Schema schema = allSchemas.get(dataset);
+
+        if (isIngraphDataset(dataset)) {
+          addIngraphDataset(dataset, schema, datasetConfig);
+        } else {
+          addPinotDataset(dataset, schema, datasetConfig);
+        }
+      }
+      refreshCaches();
+
+    } catch (IOException e) {
+      LOG.error("Exception in loading datasets", e);
+    }
+  }
+
+  public boolean isIngraphDataset(String dataset) {
```

Comment:
what if we have created table but not added any metrics? Its better to read the schema and look for some specific columns in the schema